openedx_events/bridge/avro_attrs_bridge.py

```diff
@@ -123,131 +142,72 @@ def _record_fields_for_attrs_class(
+        # This first converts data_dict to json string and then back to dict.
+
+        return json.loads(
+            json.dumps(
+                data_dict, sort_keys=True, default=self._extension_signal_serializer
+            )
         )
-        return avro_record
 
-    def serialize(self, obj) -> bytes:
+    def _extension_signal_serializer(self, value):
         """
-        Convert from attrs to a valid avro record.
+        Pass this callback into attrs.asdict function as "value_serializer" arg.
+
+        Serializes values for which an extension exists in self.extensions dict.
         """
-        avro_record = self.to_dict(obj)
-        # Try to serialize using the generated schema.
-        out = io.BytesIO()
-        fastavro.schemaless_writer(out, self.schema_dict, avro_record)
-        out.seek(0)
-        return out.read()
+        if hasattr(value, "__attrs_attrs__"):
+            return attr.asdict(value, value_serializer=self._extension_serializer)
+        for extended_class, extension in self.extensions.items():
+            if issubclass(type(value), extended_class):
+                return extension.serialize(value)
+        return value
 
     def _extension_serializer(self, _, field, value):
         """
         Pass this callback into attrs.asdict function as "value_serializer" arg.
 
-        Serializes values for which an extention exists in self.extensions dict.
+        Serialize values for which an extention exists in self.extensions dict.
         """
         extension = self.extensions.get(field.type, None)
         if extension is not None:
             return extension.serialize(value)
         return value
 
-    def deserialize(self, data: bytes, writer_schema=None) -> object:
+    def from_dict(self, data: dict):
         """
-        Deserialize data into self.attrs_cls instance.
+        Convert dict into event data that can be sent with self._signal.
 
-        Args:
-            data: bytes that you want to deserialize
-            writer_schema: pass the schema used to serialize data if it is differnt from current schema
+        :param data: Dictionary returned from AvroDeserializer
+        :return: Event data dictionary
         """
-        data_file = io.BytesIO(data)
-        if writer_schema is not None:
-            record = fastavro.schemaless_reader(
-                data_file, writer_schema, self.schema_dict
-            )
-        else:
-            record = fastavro.schemaless_reader(data_file, self.schema_dict)
-        return self.dict_to_attrs(record["data"], self._attrs_cls)
+        return dict([(top_key, self.dict_to_object(data[top_key], obj_type))
+                     for top_key, obj_type in self._signal.init_data.items()])
 
-    def dict_to_attrs(self, data: dict, attrs_cls):
+    def dict_to_object(self, data: dict, obj_type):
         """
-        Convert data into instantiated object of attrs_cls.
+        Convert dictionary entry into an instance of obj_type.
```

Comment:
1. In the class docstring, maybe it would be useful to stress that the bridge can be used to perform the following: 1) auto-generate schema: auto-generates an Avro schema based on an event definition in the form of an OpenEdxPublicSignal instance, and 2) serialize event data: serializes a specific instance of event data based on the auto-generated Avro schema.
2. For methods like this one with a vague name, it would be useful if the the docstring noted whether this was used as part of 1) schema generation, or 2) event data serialization.
3. Is this method meant to be public? Can public methods that are part of some other API note that fact? (May not apply to this method.)

Reply:
1. Would it make sense to call this `_deserialized_avro_dict_to_object`?
2. Could we add docstrings for the arguments?
3. Is `obj_type` different from the other discussion about `type_object` or `data_type`, etc. (see separate comment). Maybe we can find a consistent name, and use the docstring to clarify examples.